packages/next/build/webpack-config.ts

```diff
@@ -565,7 +565,15 @@ export default async function getBaseWebpackConfig(
   )
 
   function handleExternals(context: any, request: any, callback: any) {
-    if (request === 'next') {
+    const externalModules = [
+      'next',
+      'next/amp',
+      'next/config',
+      'next/dynamic',
+      'next/head',
+    ]
```

Comment:
Can you explain the necessity of this change? It's not clear to me what this change was made for.

We need a test case for this, and a code comment explaining it.

Reply:
It's because the paths above are symlinks that after being resolved don't respect anymore the regex  /node_modules[/\\].*\.js$/


It was explained in the commit message

```
Ensure certain 'next/*' pkgs get externalized on the server

- 'next/amp'
- 'next/config'
- 'next/dynamic'
- 'next/head'

e2e tests were failing prior to this change because e.g.

    next.js/node_modules/next/dynamic.js

is a symlink to

    next.js/packages/next/dynamic.js

So the

    if (isNextExternal || res.match(/node_modules[/\\].*\.js$/)) {

condition was not being met (the regex no longer matched).

By making sure we have a guard at the top to externalize these packages,
we avoid this issue.
```

Reply:
That bug sounds specific to our monorepo setup and not real apps. We need a solution that won't cause such a large performance overhead (`indexOf` scan for thousands of paths).

Reply:
@Timer Thoughts about the best approach here? I agree that we should avoid having code which specifically works around issues with the monorepo setup.

Looking at this code again, I'm realizing that resolving symlinks in ``res = resolveRequest(request, `${context}/`)`` results in an issue that can affect users. If you're working on a Next.js app and you've `npm link`'d a server dependency to your project's `node_modules` (e.g. `projects/my-next-app/node_modules/my-server-dep -> ../../my-server-dep`), then `my-server-dep` will no longer be externalized, because the resolved path doesn't match `res.match(/node_modules[/\\].*\.js$/)`.

I'm curious about the original motivation behind externalizing server-imported modules found in `node_modules`. Do we do this because modules might include code that is not safe to bundle with webpack, so must be executed on Node.js directly? Wondering if we can just bundle these instead.

---

I thought of an alternative approach.

To preserve most of the existing behaviour (and continue to support `npm link` as described above), I suppose we might want ``res = resolveRequest(request, `${context}/`)`` to actually tell us whether a symlink was followed from `node_modules/<dir>` to some other location. If we do this, then we could use the "original path" to know whether the module was found in `node_modules`, e.g.

```ts
res.originalPath.match(/node_modules[/\\].*\.js$/)
```

and use the "resolved path" to tell us whether `res` resolves to the same location as `baseRes`, e.g.

```ts
// Same as above: if the package, when required from the root,
// would be different from what the real resolution would use, we
// cannot externalize it.
if (baseRes.resolvedPath !== res.resolvedPath) {
  return callback()
}
```

I've implemented this here, and confirmed it works with pnpm: 02ce950e8a9cd1bac27a4145d739588b0b94bdd3

Let me know what you think.

Reply:
@Timer Can you provide some pointers on what kind of test case(s) I should add here? Should I add some unit tests for the `handleExternals` function specifically, or more of an integration-style test to make sure Next.js works with pnpm? From what I can tell, there are no existing integration tests which install packages with various package managers. I think I would need to create a new type of integration test which creates a Next.js project in a folder isolated from the monorepo's `node_modules`.